### PR TITLE
[JENKINS-26947] forcibly terminate Maven remoting channel when upstream channel get closed

### PR DESCRIPTION
**Pull request by ydubreuil**
[JENKINS-26947](http://issues.jenkins-ci.org/browse/JENKINS-26947)

Currently, when the main remoting channel is abruptly closed, Maven channel can be stuck for a while because it doesn't get notified of the disconnection.

@reviewbybees

### CODE REVIEW

**Comment by ydubreuil** [comment]
Some test are failing on CI. Had the same failure on my laptop on the master branch

**Comment by ndeloof** [comment]
@kohsuke would it maybe make sense to get this included in remoting ? Not sure channel-in-channel is a common pattern

**Comment by jglick** [comment]
> Not sure channel-in-channel is a common pattern

No, only `maven-plugin` is this crazy (except for whatever code was copied & pasted into `ivy-plugin`).

I guess I am +0. If it fixes the problem then we need it. The patch is somewhat scary.

File: src/main/java/hudson/maven/AbstractMavenProcessFactory.java [hunk 1]
```diff
@@ -234,8 +238,87 @@ private Object writeReplace() {
...
+                this.upstreamChannel = null;
+            }
+        }
+
+        public void terminateNow(IOException e) {
+            super.terminate(e);
+        }
+        
+        /**
+         * copy/paste of Channels.forProcess
```
Comment (jglick) [inline src/main/java/hudson/maven/AbstractMavenProcessFactory.java]: If there is a particular reason the existing method cannot be reused, it should be explained here.

Comment (ydubreuil) [inline src/main/java/hudson/maven/AbstractMavenProcessFactory.java]: Channel.terminate() has protected visibility so LinkedChannelCleaner.onClose() can't call it directly

Comment (ndeloof) [inline src/main/java/hudson/maven/AbstractMavenProcessFactory.java]: Did you tried to move this hack-ish class to same hudson.remoting package so you can access it's protected method ?

Comment (stephenc) [inline src/main/java/hudson/maven/AbstractMavenProcessFactory.java]: `hudson.remoting` package is in a signed jar file, so you will be SOoL if you try that

**Comment by stephenc** [comment]
:bee:

**Comment by ydubreuil** [comment]
@reviewbybees done

**Comment by ydubreuil** [comment]
@jglick Would you consider merging this PR? I can rebase to master and rerun test. I know it's pretty invasive but it's tested in real condition for some time now (there's a very big CloudBees DEV@cloud customer running fine with this patch)

**Comment by jglick** [comment]
It has a bee but it needs a merge.

**Comment by ndeloof** [comment]
@ydubreuil please rebase then we can merge

**Comment by jglick** [comment]
Rebase or just `git merge master`.

**Comment by ydubreuil** [comment]
I've just rebased to master.

**Pull request merged**